README
= YouTube

A pure Ruby object-oriented interface to the YouTube REST API documented
at http://www.youtube.com/dev.

API access requires a developer id.  You can obtain one for free at
http://www.youtube.com/my_profile_dev.

The RubyForge project is at http://rubyforge.org/projects/youtube.

The Google Group to discuss issues and bugs is at http://groups.google.com/group/ruby-youtube-library?hl=en.

== About

Implements version 2 of YouTube's API.

== Installing

Install the gem via:

  % gem install youtube

== Usage

An example as:

  require 'rubygems'
  require 'youtube'

  youtube = YouTube::Client.new 'DEVELOPER_ID'

  profile = youtube.profile('br0wnpunk')
  puts "age: " + profile.age.to_s

  favorites = youtube.favorite_videos('br0wnpunk')
  puts "number of favorite videos: " + favorites.size.to_s

  friends = youtube.friends('paolodona')
  puts "number of friends: " + friends.size.to_s
  puts "friend name: " + friends[0].user

  videos = youtube.videos_by_tag('iron maiden')
  puts "number of videos by tag iron maiden: " + videos.size.to_s

  videos = youtube.videos_by_user('whytheluckystiff')
  puts "number of videos by why: " + videos.size.to_s
  puts "title: " + videos[0].title

  videos = youtube.featured_videos
  puts "number of featured videos: " + videos.size.to_s
  puts "title: " + videos[0].title
  puts "url: " + videos[0].url
  puts "embed url: " + videos[0].embed_url
  puts "embed html: \n" + videos[0].embed_html

  details = youtube.video_details(videos[0])
  puts "detailed description: " + details.description
  puts "thumbnail url: " + details.thumbnail_url


CHANGELOG
* 2007/06/12

- [shaper] Fixed typo in VideoDetails, recoding -> recording
- [cardmagic] Changed URI.encode to CGI.escape to properly encode &
- [shaper] Test for character encoding
- [drummr77] Force XML parsed empty strings to be Strings and not Hashes
	Discovered by shaper.
- [Rob Tsuk] Added paging support to videos_by_user
- [Rob Tsuk] Created test YouTube account for running tests
- [Thomas Cox] Added support for the embed_status tag

* 2007/06/11

- [drummr77] Removed videos_by_category since it was removed by YouTube.
- [drummr77] Renamed videos_by_category_and_tag to videos_by_category_id_and_tag.
- [drummr77] Added videos_by_category_and_tag taking in a category name and tag

* 2007/02/12

- [drummr77] Added fix to check if returned video response contains an
  Array or not.  Return nil instead of erroring out.

- [drummr77] Fixed Video.length accessor.  Changed to Video.length_seconds.

- [drummr77] Added new API methods:
  youtube.videos.list_by_related (with paging),
  youtube.videos.list_by_playlist (with paging),
  youtube.videos.list_by_category (with paging)
  youtube.videos.list_by_category_and_tag (with paging)

* 2006/11/15

- [drummr77] Applied a modified version of the patch sent in by Lucas
  Carlson that uses a more functional style of programming in the API
  calls.

* 2006/11/14

- [drummr77] Changed video_details to take in a video id instead of
  the whole object.

* 2006/11/04

- [shaper] Restructured to follow standard RubyGems directory structure.

- [shaper] Created a Rakefile.

- [shaper] Created initial unit tests.

- [shaper] Fixed bug wherein responses with only a single result
  (e.g. from a videos_by_tag call) would fail to successfully parse.

- [shaper] Moved all classes within a YouTube module for cleaner namespace
  management and in particular to avoid potential conflicts.

- [shaper] Added Video.embed_html method to allow easy retrieval of HTML
  to embed the video in a web page conforming to the HTML specified on
  YouTube video pages, with width/height of video optionally specifiable.

- [shaper] Merged dirtywork.rb into the main youtube.rb file and added
  support for optionally specifying the host and api path for future
  flexibility without requiring code modifications should YouTube change
  their API access details.

- [shaper] Modified parsing response payload data to translate to the most
  appropriate Ruby objects (e.g. integers via to_i(), boolean strings to
  TrueClass/FalseClass, time strings to Time) wherever applicable.

- [shaper] Moved Video.details() into Client.video_details() for
  consistency and to avoid having to store a reference to the client in
  every Video object.

- [shaper] Updated existing and added more RDoc documentation.

- [shaper] Changed API hostname to fully-qualified www.youtube.com from
  youtube.com.

* 2006/09/28

- [drummr77] Initial Release.


test/test_api.rb
require 'rubygems'
require 'test/unit'
require 'youtube'
require 'pp'
require 'yaml'
require 'set'

# This test class assumes an active internet connection
class TestAPI < Test::Unit::TestCase
  @@DEVELOPER_API_KEY = 'RCofu-vAmeY'
  YOUTUBE_LIB_USERID = 'ytlibtest'

  def setup
    @client = YouTube::Client.new @@DEVELOPER_API_KEY
  end

  # test to ensure escaping strange characters like ampersands works
  # properly as we've been bitten by finding that they don't
  def test_videos_by_tag_escape
    videos = @client.videos_by_tag('cindy & margolis')
    _test_video_list(videos)
  end

  def test_profile
    profile = @client.profile('nutria42')

    assert_kind_of YouTube::Profile, profile
    assert_not_nil profile
    assert (profile.age >= 33)
    assert (profile.gender == 'm')
    assert (profile.country == 'US')
  end

  def test_favorite_videos
    favorites = @client.favorite_videos('br0wnpunk')

    # make sure we got some favorites
    assert_not_nil favorites
    assert (favorites.length > 0)

    # pull out one to scrutinize
    sample = favorites.first
    _test_video(sample)
    
    # user doesn't exist / test exception handling
    assert_raise RuntimeError do
      favorites = @client.favorite_videos('dsfa98sd7fa9s8d7fd8')
    end
  end

  def test_friends
    friends = @client.friends('paolodona')

    # make sure we got some friends
    assert_not_nil friends
    assert (friends.length > 0)

    # pull one out to scrutinize
    sample = friends.first
    assert_kind_of YouTube::Friend, sample

    # sanity-check some attributes to make sure we parsed properly
    assert (sample.favorite_count > 0)
    assert (sample.friend_count > 0)
    assert (sample.user && sample.user.length > 0)
  end

  def test_videos_by_tag
    # test case where videos exist
    videos = @client.videos_by_tag('iron maiden')
    _test_video_list(videos)
    videos = @client.videos_by_tag('caffe trieste')
    _test_video_list(videos)

    # test case where no videos exist
    videos = @client.videos_by_tag('u423oi4j23oi4j234io')
    assert_nil videos
  end

  def test_videos_by_user
    # test case where videos exist
    videos = @client.videos_by_user('whytheluckystiff')
    _test_video_list(videos)    
  end
  
  def test_should_return_10_videos_for_user
    videos = @client.videos_by_user(YOUTUBE_LIB_USERID)
    _test_video_list(videos)

    assert_equal 10, videos.length
    videos_seen = Set.new
    videos.each do |video|
      _validate_test_video(video, videos_seen)
    end
  end
  
  def test_should_paginate_videos_by_user
    total_videos = 10
    expected_page_size = 3
    videos_seen = Set.new
    1.upto(4) do |index|
     videos = @client.videos_by_user(YOUTUBE_LIB_USERID, index, expected_page_size)
     _test_video_list(videos)
     expected_video_count = total_videos > expected_page_size ? expected_page_size : total_videos
     assert_equal expected_video_count, videos.length
     videos.each do |video|
       _validate_test_video(video, videos_seen)
     end
     total_videos -= expected_page_size
    end    
  end

  def test_videos_by_related
    videos = @client.videos_by_related('squats')
    _test_video_list(videos)
  end
  
  def test_videos_by_playlist
    videos = @client.videos_by_playlist('4009D81B1E0277A1')
    _test_video_list(videos)
  end
  
  def test_videos_by_category_id_and_tag
    videos = @client.videos_by_category_id_and_tag(17, 'bench')
    _test_video_list(videos)
  end
  
  def test_videos_by_category_and_tag
    videos = @client.videos_by_category_and_tag(YouTube::Category::SPORTS, 'bench')
    _test_video_list(videos)
  end
  
  def test_should_not_return_any_videos_for_videos_by_category_and_tag
    videos = @client.videos_by_category_and_tag(YouTube::Category::SPORTS, 'somerandomtagthatihopedoesntexist')
    assert_nil videos
  end
  
  def test_should_get_comment_count_for_video
    videos = @client.videos_by_tag('imogen heap hide and seek')
    video = videos.sort{ |a,b| b.view_count <=> a.view_count }.first
    assert video.comment_count > 0
  end

  def test_featured_videos
    videos = @client.featured_videos
    _test_video_list(videos)
  end

  def test_video_details
    videos = @client.featured_videos
    _test_video_list(videos)

    videos.each do |video|
      details = @client.video_details(video.id)

      assert_not_nil details
      assert_kind_of YouTube::VideoDetails, details
      assert (details.author && details.author.length > 0)
      assert (details.length_seconds > 0)
      assert (details.title && details.title.length > 0)
      assert_instance_of String, details.description
      assert details.embed_allowed if details.embed_status == "ok"
      assert !details.embed_allowed if details.embed_status == "not allowed"
    end

    # make sure parameter validation is operating correctly
    assert_raise ArgumentError do
      @client.video_details(5)
    end
  end

  def test_embed_html
    videos = @client.videos_by_tag('iron maiden')
    sample_video = videos.first

    embed_html = sample_video.embed_html
    embed_url = sample_video.embed_url

    # make sure embed url is present twice in the html
    assert (_match_count(embed_url, embed_html) == 2)

    # make sure the default width and height are present
    dimension_text = "width=\"425\" height=\"350\""
    assert (_match_count(dimension_text, embed_html) == 2)

    # try changing the width and height in the embed html
    custom_embed_html = sample_video.embed_html(200, 100)
    dimension_text = "width=\"200\" height=\"100\""

    # make sure the customized width and height are present
    entries = custom_embed_html.find_all { |t| t == dimension_text }
    assert (_match_count(dimension_text, custom_embed_html) == 2)
  end

  private
    # Returns the number of times +substr+ exists within +text+.
    def _match_count (substr, text)
      return 0 if (!text || !substr)

      count = 0
      offset = 0
      while (result = text.index(substr, offset))
        count += 1
        offset = result + 1
      end
      count
    end

    def _assert_youtube_url (url)
      (url =~ /^http:\/\/www.youtube.com\//)
    end

    def _test_video_list (videos)
      # make sure we got some videos
      assert_not_nil videos
      assert (videos.length > 0)

      # make sure all video records look good
      videos.each { |video| _test_video(video) }
    end

    def _test_video (video)
      assert_kind_of YouTube::Video, video

      # sanity-check embed url to make sure it looks ok
      assert_not_nil video.embed_url
      _assert_youtube_url video.embed_url

      # check other attributes
      assert (video.thumbnail_url =~ /\.jpg$/)
      assert (video.title && video.title.length > 0)
      assert (video.length_seconds > 0)
      # make sure description came through as a string, as a single check
      # that we feel good about covering other fields since the xml parsing
      # of empty elements defaults to providing an empty hash but we want
      # a proper type everywhere.
      assert_instance_of String, video.description
      assert (video.upload_time && video.upload_time.is_a?(Time))
      _assert_youtube_url video.url
      assert (video.view_count >= 0)
      assert (video.tags && video.tags.length > 0)
      assert (video.author && video.author.length > 0)
    end
  
    def _validate_test_video(video, videos_seen = nil)
      values = YAML.load(video.description)
      index = values["index"]
    
      # make sure we only see a particular video once
      # if the particular test cares about that
      if videos_seen
        assert !videos_seen.include?(index)
        videos_seen.add(index)
      end
    
      title = sprintf("Test Video %02d", index)
      assert_equal title, video.title
      assert_equal values["length"], video.length_seconds
    
      tags = Set.new(video.tags.split)
      if index.modulo(2) == 1
        assert tags.include?("ytlodd")
      else
        assert tags.include?("ytleven")
      end
    
      if [2,3,5,7].to_set.include?(index)
        assert tags.include?("ytlprime")
      end
    end
end


examples/example.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'youtube'

youtube = YouTube::Client.new 'DEVELOPER_ID' # Get one here: <http://youtube.com/my_profile_dev>. 

profile = youtube.profile('br0wnpunk')
puts "age: " + profile.age.to_s

favorites = youtube.favorite_videos('br0wnpunk')
puts "number of favorite videos: " + favorites.size.to_s

friends = youtube.friends('paolodona')
puts "number of friends: " + friends.size.to_s
puts "friend name: " + friends.first.user

videos = youtube.videos_by_tag('iron maiden')
puts "number of videos by tag iron maiden: " + videos.size.to_s

videos = youtube.videos_by_user('whytheluckystiff')
puts "number of videos by why: " + videos.size.to_s
puts "title: " + videos.first.title

videos = youtube.featured_videos
puts "number of featured videos: " + videos.size.to_s
puts "title: " + videos.first.title
puts "url: " + videos.first.url
puts "embed url: " + videos.first.embed_url
puts "embed html: \n" + videos.first.embed_html

details = youtube.video_details(videos.first.id)
puts "detailed description: " + details.description
puts "thumbnail url: " + details.thumbnail_url


lib/youtube.rb
#--
# Copyright (c) 2006 Shane Vitarana
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#++

require 'net/http'
require 'xmlsimple'
require 'cgi'

module YouTube

  class Category
    FILMS_ANIMATION = 1
    AUTOS_VEHICLES = 2
    COMEDY = 23
    ENTERTAINMENT = 24
    MUSIC = 10
    NEWS_POLITICS = 25
    PEOPLE_BLOGS = 22
    PETS_ANIMALS = 15
    HOWTO_DIY = 26
    SPORTS = 17
    TRAVEL_PLACES = 19
    GADGETS_GAMES = 20    
  end
  
  # Main client class managing all interaction with the YouTube server.
  # Server communication is handled via method_missing() emulating an
  # RPC-like call and performing all of the work to send out the HTTP
  # request and retrieve the XML response.  Inspired by the Flickr
  # interface by Scott Raymond <http://redgreenblu.com/flickr/>.
  class Client
    # the default hostname at which the YouTube API is hosted
    DEFAULT_HOST = 'http://www.youtube.com'
    # the default api path to the YouTube API
    DEFAULT_API_PATH = '/api2_rest'
    
    def initialize(dev_id = nil, host = DEFAULT_HOST, api_path = DEFAULT_API_PATH)
      raise "developer id required" unless dev_id

      @host = host
      @api_path = api_path
      @dev_id = dev_id
    end

    # Returns a YouTube::Profile object detailing the profile information
    # regarding the supplied +username+.
    def profile(username)
      response = users_get_profile(:user => username)
      Profile.new response['user_profile']
    end

    # Returns a list of YouTube::Video objects detailing the favorite
    # videos of the supplied +username+.
    def favorite_videos(username)
      response = users_list_favorite_videos(:user => username)
      _parse_video_response(response)
    end

    # Returns a list of YouTube::Friend objects detailing the friends of
    # the supplied +username+.
    def friends(username)
      response = users_list_friends(:user => username)
      friends = response['friend_list']['friend']
      friends.is_a?(Array) ? friends.compact.map { |friend| Friend.new(friend) } : nil        
    end

    # Returns a list of YouTube::Video objects detailing the videos
    # matching the supplied +tag+.
    #
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_tag(tag, page = 1, per_page = 20)
      response = videos_list_by_tag(:tag => tag, :page => page, :per_page => per_page)
      _parse_video_response(response)
    end

    # Returns a list of YouTube::Video objects that match the
    # specified +tag+.
    #
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_related(tag, page = 1, per_page = 20)
      response = videos_list_by_related(:tag => tag, :page => page, :per_page => per_page)
      _parse_video_response(response)
    end

    # Returns a list of YouTube::Video objects with the specified
    # playlist +id+.
    #
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_playlist(id, page = 1, per_page = 20)
      response = videos_list_by_playlist(:id => id, :page => page, :per_page => per_page)
      _parse_video_response(response)
    end

    # Returns a list of YouTube::Video objects detailing the videos
    # matching the supplied category +id+ and +tag+.
    #
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_category_id_and_tag(id, tag, page = 1, per_page = 20)
      response = videos_list_by_category_and_tag(:category_id => id, :tag => tag, :page => page, :per_page => per_page)
      _parse_video_response(response)
    end

    # Returns a list of YouTube::Video objects detailing the videos
    # matching the supplied +category+ and +tag+.
    #
    # Available categories:
    #
    # * YouTube::Category::FILMS_ANIMATION
    # * YouTube::Category::AUTOS_VEHICLES
    # * YouTube::Category::COMEDY
    # * YouTube::Category::ENTERTAINMENT
    # * YouTube::Category::MUSIC
    # * YouTube::Category::NEWS_POLITICS
    # * YouTube::Category::PEOPLE_BLOGS
    # * YouTube::Category::PETS_ANIMALS
    # * YouTube::Category::HOWTO_DIY
    # * YouTube::Category::SPORTS
    # * YouTube::Category::TRAVEL_PLACES
    # * YouTube::Category::GADGETS_GAMES
    #    
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_category_and_tag(category, tag, page = 1, per_page = 20)
      videos_by_category_id_and_tag(category, tag, page, per_page)
    end

    # Returns a list of YouTube::Video objects detailing the videos
    # uploaded by the specified +username+.
    #
    # Optional parameters are:
    # +page+ = the "page" of results to retrieve (e.g. 1, 2, 3)
    # +per_page+ = the number of results per page (default: 20, max 100).
    def videos_by_user(username, page = 1, per_page = 20)
      response = videos_list_by_user(:user => username, :page => page, :per_page => per_page)
       _parse_video_response(response)
     end
    
    # Returns a list of YouTube::Video objects detailing the current
    # global set of featured videos on YouTube.
    def featured_videos
      response = videos_list_featured
      _parse_video_response(response)
    end

    # Returns a YouTube::VideoDetails object detailing additional
    # information on the supplied video id, obtained from a
    # YouTube::Video object from a previous client call.
    def video_details(video_id)
      raise ArgumentError.new("invalid video id parameter, must be string") unless video_id.is_a?(String)
      response = videos_get_details(:video_id => video_id)
      VideoDetails.new(response['video_details'])
    end

    private
      # All API methods are implemented with this method.  This method is
      # like a remote method call, it encapsulates the request/response
      # cycle to the remote host. It extracts the remote method API name
      # based on the ruby method name.
      def method_missing(method_id, *params)
        _request(method_id.to_s.sub('_', '.'), *params)
      end

      def _request(method, *params)
        url = _request_url(method, *params)
        response = XmlSimple.xml_in(_http_get(url), { 'ForceArray' => [ 'video', 'friend' ] })
        raise response['error']['description'] + " : url=#{url}" unless response['status'] == 'ok' 
        response
      end

      def _request_url(method, *params)
        param_list = String.new
        unless params.empty?
          params.first.each_pair { |k, v| param_list << "&#{k.to_s}=#{CGI.escape(v.to_s)}" }
        end
        url = "#{@host}#{@api_path}?method=youtube.#{method}&dev_id=#{@dev_id}#{param_list}"
      end

      def _http_get(url)
        Net::HTTP.get_response(URI.parse(url)).body.to_s
      end

      def _parse_video_response(response)
        videos = response['video_list']['video']
        videos.is_a?(Array) ? videos.compact.map { |video| Video.new(video) } : nil        
      end
  end

  class Friend
    attr_reader :favorite_count
    attr_reader :friend_count
    attr_reader :user
    attr_reader :video_upload_count

    def initialize(payload)
      @favorite_count = payload['favorite_count'].to_i
      @friend_count = payload['friend_count'].to_i
      @user = payload['user'].to_s
      @video_upload_count = payload['video_upload_count'].to_i
    end        
  end

  class Profile
    attr_reader :about_me
    attr_reader :age
    attr_reader :books
    attr_reader :city
    attr_reader :companies
    attr_reader :country
    attr_reader :currently_on
    attr_reader :favorite_video_count
    attr_reader :first_name
    attr_reader :friend_count
    attr_reader :gender
    attr_reader :hobbies
    attr_reader :homepage
    attr_reader :hometown
    attr_reader :last_name
    attr_reader :movies
    attr_reader :occupations
    attr_reader :relationship
    attr_reader :video_upload_count
    attr_reader :video_watch_count

    def initialize(payload)
      @about_me = payload['about_me'].to_s
      @age = payload['age'].to_i
      @books = payload['books'].to_s
      @city = payload['city'].to_s
      @companies = payload['companies'].to_s
      @country = payload['country'].to_s
      @currently_on = YouTube._string_to_boolean(payload['currently_on'])
      @favorite_video_count = payload['favorite_video_count'].to_i
      @first_name = payload['first_name'].to_s
      @friend_count = payload['friend_count'].to_i
      @gender = payload['gender'].to_s
      @hobbies = payload['hobbies'].to_s
      @homepage = payload['homepage'].to_s
      @hometown = payload['hometown'].to_s
      @last_name = payload['last_name'].to_s
      @movies = payload['movies'].to_s
      @occupations = payload['occupations'].to_s
      @relationship = payload['relationship'].to_s
      @video_upload_count = payload['video_upload_count'].to_i
      @video_watch_count = payload['video_watch_count'].to_i
    end
  end

  class Video
    attr_reader :author
    attr_reader :comment_count
    attr_reader :description
    attr_reader :embed_url
    attr_reader :id
    attr_reader :length_seconds
    attr_reader :rating_avg
    attr_reader :rating_count
    attr_reader :tags
    attr_reader :thumbnail_url
    attr_reader :title
    attr_reader :upload_time
    attr_reader :url
    attr_reader :view_count

    def initialize(payload)
      @author = payload['author'].to_s
      @comment_count = payload['comment_count'].to_i
      @description = payload['description'].to_s
      @id = payload['id']
      @length_seconds = payload['length_seconds'].to_i
      @rating_avg = payload['rating_avg'].to_f
      @rating_count = payload['rating_count'].to_i
      @tags = payload['tags']
      @thumbnail_url = payload['thumbnail_url']
      @title = payload['title'].to_s
      @upload_time = YouTube._string_to_time(payload['upload_time'])
      @url = payload['url']
      @view_count = payload['view_count'].to_i

      # the url provided via the API links to the video page -- for
      # convenience, generate the url used to embed in a page
      @embed_url = @url.delete('?').sub('=', '/')
    end    

    # Returns HTML analogous to that provided by the YouTube web site to
    # allow for easy embedding of this video in a web page.  Optional
    # +width+ and +height+ parameters allow specifying the dimensions of
    # the video for display.
    def embed_html(width = 425, height = 350)
      <<edoc
<object width="#{width}" height="#{height}">
  <param name="movie" value="#{embed_url}"></param>
  <param name="wmode" value="transparent"></param>
  <embed src="#{embed_url}" type="application/x-shockwave-flash" 
   wmode="transparent" width="#{width}" height="#{height}"></embed>
</object>
edoc
    end
  end

  class VideoDetails
    attr_reader :author
    attr_reader :channel_list
    attr_reader :comment_list
    attr_reader :description
    attr_reader :length_seconds
    attr_reader :rating_avg
    attr_reader :rating_count
    attr_reader :recording_location
    attr_reader :recording_country
    attr_reader :recording_date
    attr_reader :tags
    attr_reader :thumbnail_url
    attr_reader :title
    attr_reader :update_time
    attr_reader :upload_time
    attr_reader :view_count
    attr_reader :embed_status
    attr_reader :embed_allowed

    def initialize(payload)
      @author = payload['author'].to_s
      @channel_list = payload['channel_list']
      @comment_list = payload['comment_list']
      @description = payload['description'].to_s
      @length_seconds = payload['length_seconds'].to_i
      @rating_avg = payload['rating_avg'].to_f
      @rating_count = payload['rating_count'].to_i
      @recording_country = payload['recording_country'].to_s
      @recording_date = payload['recording_date'].to_s
      @recording_location = payload['recording_location'].to_s
      @tags = payload['tags']
      @thumbnail_url = payload['thumbnail_url']
      @title = payload['title'].to_s
      @update_time = YouTube._string_to_time(payload['update_time'])
      @upload_time = YouTube._string_to_time(payload['upload_time'])
      @view_count = payload['view_count'].to_i
      @embed_status = payload['embed_status']
      @embed_allowed = ( payload['embed_status'] == "ok" )
    end
  end

  private
    # Returns the Ruby boolean object as TrueClass or FalseClass based on
    # the supplied string value.  TrueClass is returned if the value is
    # non-nil and "true" (case-insensitive), else FalseClass is returned.
    def self._string_to_boolean(bool_str)
      (bool_str && bool_str.downcase == "true")
    end

    # Returns a Time object corresponding to the specified time string
    # representing seconds since the epoch, or nil if the string is nil.
    def self._string_to_time(time_str)
      (time_str) ? Time.at(time_str.to_i) : nil
    end    
end
